### PR TITLE
feat: implement persistance interface + index methods

### PR DESCRIPTION
**Pull request by crisog**
Initial implementation of indexer library and its postgres driver [indexer diagram](https://lucid.app/lucidchart/3a765212-5e63-41b5-ba7a-bdd4b95fee93/edit?invitationId=inv_05c6c7c9-9c80-4f2b-893c-e39262ab52c4&page=0_0#).

Missing:

- Block support
- Pagination on transactions requests

Examples of usage:

```
func DummyBlockTransactions() error {
	reqProvider := provider.NewProvider("https://dummy.com:18081/", []string{})

	driver, err := postgresdriver.NewPostgresDriverFromConnectionString("postgresql://postgres:dummy@dummy/dummy")
	if err != nil {
		return err
	}

	defer driver.Close()

	indexer := indexerLib.NewIndexer(reqProvider, driver)

	return indexer.IndexBlockTransactions(30363)
}
```

```
func DummyReadTransactions() ([]*indexerLib.Transaction, error) {
	driver, err := postgresdriver.NewPostgresDriverFromConnectionString("postgresql://postgres:dummy@dummy/dummy")
	if err != nil {
		return nil, err
	}

	defer driver.Close()

	return driver.ReadTransactions()
}
```

```
func DummyReadTransaction() (*indexerLib.Transaction, error) {
	driver, err := postgresdriver.NewPostgresDriverFromConnectionString("postgresql://postgres:dummy@dummy/dummy")
	if err != nil {
		return nil, err
	}

	defer driver.Close()

	return driver.ReadTransaction("AF5BB3EAFF431E2E5E784D639825979FF20A779725BFE61D4521340F70C3996D0")
}
```

### CODE REVIEW

File: indexer.go [hunk 1]
```diff
@@ -1,31 +1,71 @@
...
 
+import (
+	"github.com/pokt-foundation/pocket-indexer/types"
+)
+
 // Provider interface of needed provider functions
 type Provider interface {
 	GetBlockHeight() (int, error)
-	// TODO: add missing functions
+	GetBlock(blockHeight int) (*types.Block, error)
```
Comment (pablocampogo) [inline indexer.go:10]: These should return the values they return from pocket go, that's what the provider will return

File: types/transaction.go [hunk 2]
```diff
@@ -0,0 +1,17 @@
+package types
+
+import (
+	"math/big"
+)
+
+type Transaction struct {
```
Comment (pablocampogo) [inline types/transaction.go:7]: I guess the fields on this struct should be public (start capital letters)

File: types/block.go [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package types
+
+import (
+	"time"
+)
+
+type Block struct {
```
Comment (pablocampogo) [inline types/block.go:7]: Do we need a stuct for these?

**Review by pablocampogo — CHANGES_REQUESTED** [review]

File: indexer.go [hunk 4]
```diff
@@ -1,31 +1,115 @@
...
+
+		tx := &types.Transaction{
+			Status:   status,
+			Hash:     transaction.Hash,
+			Sender:   transaction.TxResult.Signer,
+			Receiver: transaction.TxResult.Recipient,
+			Height:   transaction.Height,
+			Code:     transaction.TxResult.Code,
+			Fee:      *big.NewInt(int64(fee)),
+			Amount:   *big.NewInt(int64(0)),
```
Comment (crisog) [inline indexer.go:101]: This 0 is a placeholder - it needs some manipulation due to the variety of transaction types

File: indexer.go [hunk 5]
```diff
@@ -1,31 +1,77 @@
...
-// Writer interface of write function
-type Writer interface {
-	// TODO: add missing functions
+// Persistance layer interface (database, in-memory, etc.)
+type Persistance interface {
+	// Writes
+	WriteBlock(block *provider.GetBlockOutput) error
+	WriteTransactions(txs []*provider.Transaction) error
+	// Reads
+	ReadBlock(blockHeight int) (interface{}, error)
```
Comment (pablocampogo) [inline indexer.go:20]: I think we should enforce `map[string]interface{}` since all these will be key-value pairs

File: indexer.go [hunk 6]
```diff
@@ -1,31 +1,77 @@
...
+		return err
+	}
+
+	txs := make([]*provider.Transaction, 0)
+
+	for _, transaction := range blockTransactionsOutput.Txs {
+		txs = append(txs, transaction)
+	}
+
+	writeErr := i.persistance.WriteTransactions(txs)
```
Comment (pablocampogo) [inline indexer.go:70]: You can send here `blockTransactionsOutput.Txs` directly

**Review by pablocampogo — CHANGES_REQUESTED** [review]

File: indexer.go [hunk 7]
```diff
@@ -1,31 +1,77 @@
...
-	// TODO: add missing functions
+	GetBlock(blockHeight int) (*provider.GetBlockOutput, error)
+	GetBlockTransactions(blockHeight int) (*provider.GetBlockTransactionsOutput, error)
 }
 
-// Writer interface of write function
-type Writer interface {
-	// TODO: add missing functions
+// Persistance layer interface (database, in-memory, etc.)
+type Persistance interface {
```
Comment (Evalir) [inline indexer.go:15]:
```suggestion
type Persistence interface {
```

File: indexer.go [hunk 8]
```diff
@@ -1,31 +1,77 @@
...
 	GetBlockHeight() (int, error)
-	// TODO: add missing functions
+	GetBlock(blockHeight int) (*provider.GetBlockOutput, error)
+	GetBlockTransactions(blockHeight int) (*provider.GetBlockTransactionsOutput, error)
 }
 
-// Writer interface of write function
-type Writer interface {
-	// TODO: add missing functions
+// Persistance layer interface (database, in-memory, etc.)
```
Comment (Evalir) [inline indexer.go:14]:
```suggestion
// Persistence layer interface (database, in-memory, etc.)
```

File: indexer.go [hunk 9]
```diff
@@ -1,31 +1,77 @@
...
+	ReadTransaction(hash string) (interface{}, error)
+	ReadBlockTransactions(blockHeight int) ([]interface{}, error)
 }
 
 // Indexer struc handler for Indexer functions
 type Indexer struct {
-	provider Provider
-	writer   Writer
+	provider    Provider
+	persistance Persistance
```
Comment (Evalir) [inline indexer.go:28]:
```suggestion
	persistence Persistence
```

File: indexer.go [hunk 10]
```diff
@@ -1,31 +1,77 @@
...
 }
 
 // NewIndexer returns Indexer instance with given input
-func NewIndexer(provider Provider, writer Writer) *Indexer {
+func NewIndexer(provider Provider, persistance Persistance) *Indexer {
```
Comment (Evalir) [inline indexer.go:32]:
```suggestion
func NewIndexer(provider Provider, persistence Persistence) *Indexer {
```

File: indexer.go [hunk 11]
```diff
@@ -1,31 +1,77 @@
...
 	return &Indexer{
-		provider: provider,
-		writer:   writer,
+		provider:    provider,
+		persistance: persistance,
```
Comment (Evalir) [inline indexer.go:35]:
```suggestion
		persistence: persistence,
```

File: indexer.go [hunk 12]
```diff
@@ -1,31 +1,77 @@
...
-// IndexBlock transfers all transactions to a known strcuture and saves them
+// IndexBlock converts block details to a known structure and saves them
 func (i *Indexer) IndexBlock(blockHeight int) error {
+	blockOutput, err := i.provider.GetBlock(blockHeight)
+
+	if err != nil {
+		return err
+	}
+
+	writeErr := i.persistance.WriteBlock(blockOutput)
```
Comment (Evalir) [inline indexer.go:47]:
```suggestion
	writeErr := i.persistence.WriteBlock(blockOutput)
```

**Review by Evalir — CHANGES_REQUESTED** [review]
Typos :)

File: indexer.go (hunk 7 above)
Comment (crisog) [reply indexer.go]: lol didn't realize the typo even tho i looked at it for hours

File: indexer.go [hunk 13]
```diff
@@ -1,14 +1,21 @@
...
 
-// Writer interface of write function
+// Writer interface for write methods to index
 type Writer interface {
-	// TODO: add missing functions
+	WriteBlock(block *Block) error
+	WriteTransactions(txs []*Transaction) error
 }
 
 // Indexer struc handler for Indexer functions
```
Comment (Evalir) [inline indexer.go:21]:
```suggestion
// Indexer struct handler for Indexer functions
```

File: indexer.go [hunk 14]
```diff
@@ -25,7 +32,117 @@ func NewIndexer(provider Provider, writer Writer) *Indexer {
...
+	Proof           *provider.TransactionProof
+	StdTx           *provider.StdTx
+	TxResult        *provider.TxResult
+	Tx              string
+	Entropy         int
+	Fee             int
+	FeeDenomination string
+}
+
+func convertProvTransactionToTransaction(provTransaction *provider.Transaction) *Transaction {
```
Comment (Evalir) [inline indexer.go:55]: What is Prov? I can't really figure out what it means, so I think we should add some docs or pick a different function name.

File: indexer.go [hunk 15]
```diff
@@ -25,7 +32,117 @@ func NewIndexer(provider Provider, writer Writer) *Indexer {
...
+	// TODO: add pagination support
+	blockTransactionsOutput, err := i.provider.GetBlockTransactions(blockHeight, nil)
+	if err != nil {
+		return err
+	}
+
+	var transactions []*Transaction
+
+	for _, tx := range blockTransactionsOutput.Txs {
+		transactions = append(transactions, convertProvTransactionToTransaction(tx))
```
Comment (Evalir) [inline indexer.go:115]: Ah, so it's trying to marshal transactions. Why not `marshallTransaction` or `tryMarshallTransaction`?

File: indexer.go [hunk 16]
```diff
@@ -25,7 +32,117 @@ func NewIndexer(provider Provider, writer Writer) *Indexer {
...
+
+	err = i.writer.WriteTransactions(transactions)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Block struct handler of all block field to be indexed
```
Comment (Evalir) [inline indexer.go:126]:
```suggestion
// Block struct handler of all block fields to be indexed
```

File: postgres-driver/postgres_driver.go [hunk 17]
```diff
@@ -0,0 +1,150 @@
...
+
+// NewPostgresDriverFromSQLDBInstance returns PostgresDriver instance from sdl.DB instance
+// mostly used for mocking tests
+func NewPostgresDriverFromSQLDBInstance(db *sql.DB) *PostgresDriver {
+	return &PostgresDriver{
+		DB: sqlx.NewDb(db, "postgres"),
+	}
+}
+
+type transaction struct {
```
Comment (Evalir) [inline postgres-driver/postgres_driver.go:44]: Don't we have the same struct in another file? Why resort to duplication?

Comment (Evalir) [reply postgres-driver/postgres_driver.go:44]: Right, I see that this is for the internal state in the DB. This should be clear in naming and in inline docs ideally.

**Review by Evalir — CHANGES_REQUESTED** [review]
So far so good, just a few comments/things to change:

File: indexer.go (hunk 15 above)
Comment (pablocampogo) [reply indexer.go:115]: In Go, marshal is most of the time related to JSON not in between strutcts

**Review by rem1niscence — APPROVED** [review]
Excellent, beautiful, perfect

**Review by pablocampogo — APPROVED** [review]
Who wrote this amazing piece of code?

**Pull request merged**